crates/matrix-sdk/src/room/mod.rs

```diff
@@ -3241,6 +3246,17 @@ impl Room {
             };
         Ok(current_requests_to_join_ids)
     }
+
+    /// Subscribes to the set of requests to join that have been marked as
+    /// 'seen'.
+    pub async fn subscribe_to_seen_requests_to_join_ids(
+        &self,
+    ) -> Result<(HashSet<OwnedEventId>, impl Stream<Item = HashSet<OwnedEventId>>)> {
+        let current = self.get_seen_requests_to_join_ids().await?;
+        let subscriber =
+            self.seen_requests_to_join_ids.subscribe().map(|values| values.unwrap_or_default());
+        Ok((current, subscriber))
+    }
```

Comment:
Unfortunately, this is false observability. `seen_requests_to_join_ids` exists in `Room`, which is created on-demand, e.g. in `Client::get_room()`, `Client::rooms()`. Given a `RoomId`, there might be multiple, different `Room` objects alive at the same time, all representing the room with that id.

As a result, if you do the following:
- create `Room` object for room id XYZ
- create a second `Room` object for the same room id
- call this method ^ on the second object
- set a seen_requests_to_join on the first object

Then the second Room object will NOT emit an update about the seen_requests_to_join emitted from the first object.

Solving this requires getting back to the drawing board, and figuring a better place where to put the `seen_requests_to_join_ids` set:

- the `BaseRoom` aka `matrix_sdk_base::Room` would be a good candidate, as all the data it contains always links to the same identity. (It'd be worth having some kind of `BaseRoomInner` to make this clearer, but oh well, so many wrappers already.) So it'd be possible to include an `Arc` in there for the set, and modify it from all the `BaseRoom`, then `Room` instances.
- an alternative is to put this into `RoomInfo`, which is already unique per room + observable :eyes: If you've included this set in the `RoomInfo` in a subsequent patch, it might be worth doing it sooner rather than later. Otherwise, putting it in `BaseRoom` might be sufficient.

Reply:
> the BaseRoom aka matrix_sdk_base::Room would be a good candidate

I might be wrong, but isn't [there](https://github.com/matrix-org/matrix-rust-sdk/pull/4338/files#diff-07c79eee28a9bd6d7e2c163f4a1b1a9654da0065698126793adc636f6fa846e3R155) where it is at the moment? I think it just gives you the impression that it's in `sdk::Room` because it's using the `Defer` impl to access `BaseRoom`. 

I have another question though: since this data lives in the store, the field gets populated the first time `Room::get_seen_join_requests` is called. Maybe there's a better way to handle this?

Reply:
> I might be wrong, but isn't [there](https://github.com/matrix-org/matrix-rust-sdk/pull/4338/files#diff-07c79eee28a9bd6d7e2c163f4a1b1a9654da0065698126793adc636f6fa846e3R155) where it is at the moment? I think it just gives you the impression that it's in sdk::Room because it's using the Defer impl to access BaseRoom.

oh my, absolute facepalm of a review, please ignore me then :'D

> I have another question though: since this data lives in the store, the field gets populated the first time Room::get_seen_join_requests is called. Maybe there's a better way to handle this?

It's fine, I get it that it's an in-memory cache for data that lives in the store, so there's no way to get it sooner than that anyways (and it's good that it's filled lazily on demand too).